R/expr__string.R

```diff
@@ -591,6 +614,9 @@ ExprStr_count_match = function(pattern) {
 #' @examples
 #' df = pl$DataFrame(s = c("foo bar", "foo-bar", "foo bar baz"))
 #' df$select(pl$col("s")$str$split(by = " "))
+#'
+#' df = pl$DataFrame(s = c("foo^bar", "foo_bar", "foo*bar*baz"), "by" = c("_", "_", "*"))
+#' df$select(pl$col("s")$str$split(by = pl$col("by"))$alias("split"))
 ExprStr_split = function(by, inclusive = FALSE) {
   unwrap(
     .pr$Expr$str_split(self, result(by), result(inclusive)),
```

Comment:
@sorhawell I didn't have to change anything for `str$split()` to accept expressions but I'm confused about why `result()` is used.

Across the codebase, when we deal with inputs using expressions, I've seen use of `wrap_e()` (for example in `ExprStr_extract_all`), `wrap_e_result()` (for example in `ExprStr_replace`) and `result()` (like here) so I don't know which one should be used. Sorry if you already explained it to me N times but I'm really missing something here

Reply:
`result()` is the reverse of `unwrap()`. It is basically a tryCatch. We should never ever* use plain stop() or any internal function that raises errors for user errors. The user has little chance to know where the error occurred in the query and will have spend time debugging every arg in every method and start hating on polars.

One exception is internal errors and failed run-time assertions. They should scream INTERNAL ERROR and make any user immediately complain to us in an issue :) Then a plain stop("Internal Error: ...") is fine, it is like a rust panic!.

If we cannot/will not prevent some internal function from raising an error, we can just wrap in `result()` to convert it to a Result which on R side is either list(ok = some_ok_value, err = NULL) or list(ok=NULL, err = some_err_value). result() also helps to upgrade any R err condition to our error type RPolarsError. Sometimes there is a whole block of code that does a lot of R stuff and who knows what error it could throw. Just wrap the entire block in a `result()`

`wrap_e()` is very similar to py-polars which also have internal functions to convert almost anything to an Expr.
`wrap_e()` is almost never needed anymore because we have `robj_to!(Expr, some_arg)?` on rust-side. It is rarely needed when defining some binary operators, otherwise deprecated. The problem with `wrap_e()` is that it will throw an error immediately if conversion failed. Then the dev could have written `result(wrap_e(x))` and is just short-hand for that `wrap_e_result()`.

The polars project is very big. For a year I knew I had to translate x methods a day to get close to feature complete and that I would learn better patterns along the way. Whenever fixing some method I would also rewrite it with `robj_to!`. 

extendr internal conversions are far from great and cannot be customized, without introducing features which will likely never be accepted as PR in extendr. robj_to! is the best way known to me get very concise error handling and custom type conversion. The alternative is generic traits but finally the syntax is worse, it is less flexible than a macro_rule and more mind bending.

Reply:
This was an extreme take on catching ALL errors.  I think I went too far catching all errors of args also. I did continue that.

```R
ExprStr_split = function(by, inclusive = FALSE) {
  unwrap(
    .pr$Expr$str_split(self, result(by), result(inclusive)),
    context = "in str$split:"
  )
}
```
example of how even a literal error in the arg is caught and added the context of the method
```R
> pl$lit("hey mom")$str$split(stop("oups"))$lit_to_s()
Error: in str$split: Execution halted with the following contexts
1: The argument [by] caused an error
2: oups
When calling: pl$lit("hey mom")$str$split(stop("oups"))
```

example of pl$lit which will not catch literal errors in args
```R
> pl$lit(stop(oups))
Error: object 'oups' not found
```

I would to day just write

```R
ExprStr_split = function(by, inclusive = FALSE) {
  unwrap(
    .pr$Expr$str_split(self, by, inclusive),
    context = "in str$split:"
  )
}
```

Reply:
Thanks once again for the detailed explanation. I opened https://github.com/pola-rs/r-polars/issues/422 to track this since it doesn't prevent the merge here.

> The polars project is very big. For a year I knew I had to translate x methods a day to get close to feature complete and that I would learn better patterns along the way.

That completely makes sense and I didn't mean any criticism in my comment. I simply wanted to point out that it would be worth harmonizing the internal expression handling since most features are now implemented ;)